main.tf

```diff
@@ -1,18 +1,16 @@
 resource "aws_ram_resource_share" "subnet_sharing" {
-  count                     = length(var.private_subnet_arns) > 0 || length(var.public_subnet_arns) > 0 ? 1 : 0
+  count                     = length(var.private_subnet_arns) >= 0 || length(var.public_subnet_arns) >= 0 ? 1 : 0
   name                      = "${var.prepend_resource_type ? "resource-share-" : ""}subnets-${var.name}"
   allow_external_principals = true
-  tags                      = var.tags
+  tags                      = merge({ resource-type = "ec2:Subnet" }, var.tags)
 }
-
 resource "aws_ram_resource_association" "private_subnets" {
-  for_each           = toset(var.private_subnet_arns)
-  resource_arn       = each.value
+  count              = length(var.private_subnet_arns)
```

Comment:
Are you very very sure that `count` doesn't give you the exact same error for a new deployment? As I believe it requires the same deterministic things the `for_each` requires...

Reply:
Yeap, tested numerous times. That is basically why the initial MR was with count and was thoroughly tested. Apparently the count gets calculated in the graph later in the graph than for_each and can be used for dynamically created resources.

Reply:
I just revalidated as well, my tests below

```
module "example_vpc_for_each" {
  source                     = "git@github.com:schubergphilis/terraform-aws-mcaf-vpc.git?ref=v1.4.0"
  name                       = "example"
  cidr_block                 = "10.112.0.0/19"
  availability_zones         = local.availability_zones
  tags                       = { environment = "example" }
  public_subnet_bits         = 22
  private_subnet_bits        = 22
  prepend_resource_type      = true
  private_s3_endpoint        = true
  share_private_subnets      = true
  share_public_subnets       = true
}

module "example_vpc_count" {
  source                     = "git@github.com:schubergphilis/terraform-aws-mcaf-vpc.git?ref=v1.4.0"
  name                       = "example"
  cidr_block                 = "10.114.0.0/19"
  availability_zones         = local.availability_zones
  tags                       = { environment = "example" }
  public_subnet_bits         = 22
  private_subnet_bits        = 22
  prepend_resource_type      = true
  private_s3_endpoint        = true
  share_private_subnets      = true
  share_public_subnets       = true
}
```

doing above and locally changing the subnet sharing module in the vpc count example produced the following

```
terraform plan -target=module.example_vpc_for_each
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

module.example_vpc_for_each.data.aws_region.current: Refreshing state...
module.example_vpc_for_each.data.aws_iam_policy_document.log_stream_action: Refreshing state...

------------------------------------------------------------------------

Warning: Resource targeting is in effect

You are creating a plan with the -target option, which means that the result
of this plan may not represent all of the changes requested by the current
configuration.

The -target option is not for routine use, and is provided only for
exceptional situations such as recovering from errors or mistakes, or when
Terraform specifically suggests to use it as part of an error message.


Error: Invalid for_each argument

  on .terraform/modules/example_vpc_for_each.subnet_sharing/main.tf line 9, in resource "aws_ram_resource_association" "private_subnets":
   9:   for_each           = toset(var.private_subnet_arns)

The "for_each" value depends on resource attributes that cannot be determined
until apply, so Terraform cannot predict how many instances will be created.
To work around this, use the -target argument to first apply only the
resources that the for_each depends on.


Error: Invalid for_each argument

  on .terraform/modules/example_vpc_for_each.subnet_sharing/main.tf line 15, in resource "aws_ram_resource_association" "public_subnets":
  15:   for_each           = toset(var.public_subnet_arns)

The "for_each" value depends on resource attributes that cannot be determined
until apply, so Terraform cannot predict how many instances will be created.
To work around this, use the -target argument to first apply only the
resources that the for_each depends on.
```

and doing it targetted against count 

```
terraform plan -target=module.example_vpc_count
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

module.example_vpc_count.data.aws_region.current: Refreshing state...
module.example_vpc_count.data.aws_iam_policy_document.log_stream_action: Refreshing state...

------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # module.example_vpc_count.aws_eip.nat[0] will be created
  + resource "aws_eip" "nat" {
      + allocation_id     = (known after apply)
+ LOTS MORE RESULTS 
```

for completeness sake I also run a full plan without targetting with the same result.